# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webrtc/libwebrtcglue/VideoConduit.h
# Commit: 88a9fe04aaa5
# Full Hash: 88a9fe04aaa582676a84734d67c9069b68a7f3e1
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2021-11-01 18:17:33
# Regressor Bug: 1654112
# File Overlap Count: 5
# Description:
#   Bug 1654112 - Update Audio- and VideoConduit stats API and integration. r=ng,bwc
#   
#   For AudioConduit this patch:
#   - Uses correct threads per upstream, pending changes due to the threading model
#     update that's coming
# ==============================================================================

diff -r 3eae9f9279f6 -r 88a9fe04aaa5 dom/media/webrtc/libwebrtcglue/VideoConduit.h
--- a/dom/media/webrtc/libwebrtcglue/VideoConduit.h	Fri Jan 29 15:38:54 2021 +0100
+++ b/dom/media/webrtc/libwebrtcglue/VideoConduit.h	Mon Jan 18 15:28:29 2021 +0100
@@ -252,32 +252,9 @@
 
   bool SetRemoteSSRCLocked(uint32_t ssrc, uint32_t rtxSsrc);
 
-  bool GetSendPacketTypeStats(
-      webrtc::RtcpPacketTypeCounter* aPacketCounts) override;
-
-  bool GetRecvPacketTypeStats(
-      webrtc::RtcpPacketTypeCounter* aPacketCounts) override;
-
-  void PollStats();
-  void UpdateVideoStatsTimer();
-  bool GetVideoEncoderStats(double* framerateMean, double* framerateStdDev,
-                            double* bitrateMean, double* bitrateStdDev,
-                            uint32_t* droppedFrames, uint32_t* framesEncoded,
-                            Maybe<uint64_t>* qpSum) override;
-  bool GetVideoDecoderStats(double* framerateMean, double* framerateStdDev,
-                            double* bitrateMean, double* bitrateStdDev,
-                            uint32_t* discardedPackets,
-                            uint32_t* framesDecoded) override;
-  bool GetRTPReceiverStats(unsigned int* jitterMs,
-                           unsigned int* cumulativeLost) override;
-  bool GetRTCPReceiverReport(uint32_t* jitterMs, uint32_t* packetsReceived,
-                             uint64_t* bytesReceived, uint32_t* cumulativeLost,
-                             Maybe<double>* aOutRttSec) override;
-  bool GetRTCPSenderReport(unsigned int* packetsSent, uint64_t* bytesSent,
-                           DOMHighResTimeStamp* aRemoteTimestamp) override;
-
-  Maybe<mozilla::dom::RTCBandwidthEstimationInternal> GetBandwidthEstimation()
-      override;
+  Maybe<webrtc::VideoReceiveStream::Stats> GetReceiverStats() const override;
+  Maybe<webrtc::VideoSendStream::Stats> GetSenderStats() const override;
+  webrtc::Call::Stats GetCallStats() const override;
 
   void GetRtpSources(nsTArray<dom::RTCRtpSourceEntry>& outSources) override;
   bool AddFrameHistory(dom::Sequence<dom::RTCVideoFrameHistoryInternal>*
@@ -290,15 +267,7 @@
     mAllowSsrcChange = false;
   }
 
-  Maybe<RefPtr<VideoSessionConduit>> AsVideoSessionConduit() override {
-    return Some(RefPtr<VideoSessionConduit>(this));
-  }
-
-  void RecordTelemetry() const override {
-    ASSERT_ON_THREAD(mStsThread);
-    mSendStreamStats.RecordTelemetry();
-    mRecvStreamStats.RecordTelemetry();
-  }
+  void RecordTelemetry() override;
 
   void SetRtcpEventObserver(mozilla::RtcpEventObserver* observer) override;
 
@@ -307,145 +276,6 @@
   WebrtcVideoConduit(const WebrtcVideoConduit&) = delete;
   void operator=(const WebrtcVideoConduit&) = delete;
 
-  /**
-   * Statistics for the Call associated with this VideoConduit.
-   * Single threaded.
-   */
-  class CallStatistics {
-   public:
-    explicit CallStatistics(nsCOMPtr<nsISerialEventTarget> aStatsThread)
-        : mStatsThread(aStatsThread) {}
-    void Update(const webrtc::Call::Stats& aStats);
-    Maybe<mozilla::dom::RTCBandwidthEstimationInternal> Stats() const;
-    Maybe<DOMHighResTimeStamp> RttSec() const;
-
-   protected:
-    const nsCOMPtr<nsISerialEventTarget> mStatsThread;
-
-   private:
-    Maybe<webrtc::Call::Stats> mStats = Nothing();
-    Maybe<DOMHighResTimeStamp> mRttSec = Nothing();
-  };
-
-  /**
-   * Shared statistics for receive and transmit video streams.
-   * Single threaded.
-   */
-  class StreamStatistics {
-   public:
-    explicit StreamStatistics(nsCOMPtr<nsISerialEventTarget> aStatsThread)
-        : mStatsThread(aStatsThread) {}
-    void Update(const double aFrameRate, const double aBitrate,
-                const webrtc::RtcpPacketTypeCounter& aPacketCounts);
-    /**
-     * Returns gathered stream statistics
-     * @param aOutFrMean: mean framerate
-     * @param aOutFrStdDev: standard deviation of framerate
-     * @param aOutBrMean: mean bitrate
-     * @param aOutBrStdDev: standard deviation of bitrate
-     */
-    bool GetVideoStreamStats(double& aOutFrMean, double& aOutFrStdDev,
-                             double& aOutBrMean, double& aOutBrStdDev) const;
-
-    /**
-     * Accumulates video quality telemetry
-     */
-    void RecordTelemetry() const;
-    const webrtc::RtcpPacketTypeCounter& PacketCounts() const;
-    bool Active() const;
-    void SetActive(bool aActive);
-    virtual bool IsSend() const { return false; };
-
-   protected:
-    const nsCOMPtr<nsISerialEventTarget> mStatsThread;
-
-   private:
-    bool mActive = false;
-    RunningStat mFrameRate;
-    RunningStat mBitRate;
-    webrtc::RtcpPacketTypeCounter mPacketCounts;
-  };
-
-  /**
-   * Statistics for sending streams. Single threaded.
-   */
-  class SendStreamStatistics : public StreamStatistics {
-   public:
-    explicit SendStreamStatistics(nsCOMPtr<nsISerialEventTarget> aStatsThread)
-        : StreamStatistics(
-              std::forward<nsCOMPtr<nsISerialEventTarget>>(aStatsThread)) {}
-    /**
-     * Returns the calculate number of dropped frames
-     */
-    uint32_t DroppedFrames() const;
-    /**
-     * Returns the number of frames that have been encoded so far
-     */
-    uint32_t FramesEncoded() const;
-    void Update(const webrtc::VideoSendStream::Stats& aStats,
-                uint32_t aConfiguredSsrc);
-    /**
-     * Call once for every frame delivered for encoding
-     */
-    void FrameDeliveredToEncoder();
-
-    bool SsrcFound() const;
-    uint32_t JitterMs() const;
-    uint32_t PacketsLost() const;
-    uint64_t BytesReceived() const;
-    uint32_t PacketsReceived() const;
-    Maybe<uint64_t> QpSum() const;
-    bool IsSend() const override { return true; };
-
-   private:
-    uint32_t mDroppedFrames = 0;
-    uint32_t mFramesEncoded = 0;
-    int32_t mFramesDeliveredToEncoder = 0;
-
-    bool mSsrcFound = false;
-    uint32_t mJitterMs = 0;
-    uint32_t mPacketsLost = 0;
-    uint64_t mBytesReceived = 0;
-    uint32_t mPacketsReceived = 0;
-    Maybe<uint64_t> mQpSum;
-  };
-
-  /**
-   * Statistics for receiving streams. Single threaded.
-   */
-  class ReceiveStreamStatistics : public StreamStatistics {
-   public:
-    explicit ReceiveStreamStatistics(
-        nsCOMPtr<nsISerialEventTarget> aStatsThread)
-        : StreamStatistics(
-              std::forward<nsCOMPtr<nsISerialEventTarget>>(aStatsThread)) {}
-    uint32_t BytesSent() const;
-    /**
-     * Returns the number of discarded packets
-     */
-    uint32_t DiscardedPackets() const;
-    /**
-     * Returns the number of frames decoded
-     */
-    uint32_t FramesDecoded() const;
-    uint32_t JitterMs() const;
-    uint32_t PacketsLost() const;
-    uint32_t PacketsSent() const;
-    uint32_t Ssrc() const;
-    DOMHighResTimeStamp RemoteTimestamp() const;
-    void Update(const webrtc::VideoReceiveStream::Stats& aStats);
-
-   private:
-    uint32_t mBytesSent = 0;
-    uint32_t mDiscardedPackets = 0;
-    uint32_t mFramesDecoded = 0;
-    uint32_t mJitterMs = 0;
-    uint32_t mPacketsLost = 0;
-    uint32_t mPacketsSent = 0;
-    uint32_t mSsrc = 0;
-    DOMHighResTimeStamp mRemoteTimestamp = 0;
-  };
-
   // Utility function to dump recv codec database
   void DumpCodecDB() const;
 
@@ -531,14 +361,11 @@
   bool mUpdateResolution = false;
   int mSinkWantsPixelCount = std::numeric_limits<int>::max();
 
-  // Bookkeeping of send stream stats. Sts thread only.
-  SendStreamStatistics mSendStreamStats;
-
-  // Bookkeeping of send stream stats. Sts thread only.
-  ReceiveStreamStatistics mRecvStreamStats;
-
-  // Bookkeeping of call stats. Sts thread only.
-  CallStatistics mCallStats;
+  // Bookkeeping of stats for telemetry. Sts thread only.
+  RunningStat mSendFramerate;
+  RunningStat mSendBitrate;
+  RunningStat mRecvFramerate;
+  RunningStat mRecvBitrate;
 
   // Must call webrtc::Call::DestroyVideoReceive/SendStream to delete this.
   // Written only on main thread. Guarded by mMutex, except for reads on main.
@@ -644,11 +471,6 @@
   // Main thread only
   uint64_t mRecvCodecPluginID = 0;
 
-  // Timer that updates video stats periodically. Main thread only.
-  nsCOMPtr<nsITimer> mVideoStatsTimer;
-  // True if mVideoStatsTimer is running. Main thread only.
-  bool mVideoStatsTimerActive = false;
-
   // Main thread only
   std::string mPCHandle;
 
